### PR TITLE
Update and refactor sounds

### LINKED ISSUE #1516: Wrong audio player stopped when playing multiple Sounds with play_concurrently()
When `play_concurrently()` is called on a Sound:
- a new HTML audio element is created and assigned to `_safeaudio` in `externalLibs/sound/sounds.js`,
- an event listener that calls `stop()` when the `ended` event is fired is attached to that audio element, and
- the audio is played.

However, `stop()` doesn't stop the correct audio player if another Sound is played while the previous Sound(s) are still playing, because it stops playback on the audio element assigned to `_safeaudio`, i.e. the last Sound that was played.

### Steps to reproduce:
(this may only work on the quest "The Magical Tone Matrix")
```
play_concurrently(sine_sound(500, 1));
play_concurrently(sine_sound(1000, 5));
```

The 1000Hz tone ends up being played for only a fraction of a second longer than the 500Hz tone, instead of being played for 5s.

### Temporary workaround:
```
play_concurrently(sound_to_play);
play_concurrently(silence_sound(0));
```

This causes a "dummy" audio element to be assigned to `_safeaudio` so that other currently playing Sounds won't be affected by `stop()`

### PR DESCRIPTION
**Pull request by Prearo**
### Description

- Refactor sound playback functions to use WebAudio API
- Remove old audio encoder
- Remove old and/or duplicated code
- Refactor existing code for clarity and to use ES6 syntax
- Fixes #1516 

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [x] Code quality improvements

### How to test

Use the sounds library to play sounds.
```javascript
const test1 = list(piano(60, 2), bell(61, 2), violin(62, 2));
play(consecutively(test1));
```

```javascript
const test2 = consecutively(list(piano(60, 2), piano(64, 2)));
const test3 = consecutively(list(silence_sound(1), violin(48, 3))); 

play(test2);
play_concurrently(test3);
```

### Checklist

- [x] I have tested this code

### CODE REVIEW

**Comment by Prearo** [comment]
Accidentally closed the pr.

File: public/externalLibs/sound/soundToneMatrix.js [hunk 1]
```diff
@@ -417,7 +417,8 @@ function letter_name_to_frequency(note) {
 }
 
 function midi_note_to_frequency(note) {
-  return 8.1757989156 * Math.pow(2, (note / 12));
+  // Frequency of midi note 0 / C-1.
+  return 8.17579891564371 * Math.pow(2, (note / 12));
```
Comment (martin-henz) [inline public/externalLibs/sound/soundToneMatrix.js:421]: Give a name to this value and declare it at a prominent place: something like
const frequency_of_c0 = ...;

File: public/externalLibs/sound/sounds.js [hunk 2]
```diff
@@ -576,15 +549,45 @@ function sawtooth_sound(freq, duration) {
...
+        // Create mono buffer
+        let theBuffer = _audioplayer.createBuffer(1, FS * get_duration(sound), FS);
+        let channel = theBuffer.getChannelData(0);
+
+        // Discretize the function and clip amplitude
+        let temp;
+        const wave = get_wave(sound);
+        for (let i = 0; i < channel.length; i++) {
+            temp = wave(i/FS);
+            channel[i] = temp > 1 ? 1 : temp < -1 ? -1 : temp;
```
Comment (Jefferson111) [inline public/externalLibs/sound/sounds.js:576]: I see that all of the `raw_to_audio` functionality except the clipping portion are removed
`channel[i] = temp > 1 ? 1 : temp < -1 ? -1 : temp;`
The quantization and exponential filter aspects from `raw_to_audio` are removed, is this change intended?

File: public/externalLibs/sound/sounds.js [hunk 3]
```diff
@@ -245,63 +205,84 @@ function play_unsafe(sound) {
...
```
Comment (Jefferson111) [inline public/externalLibs/sound/sounds.js:262]: Same here for the `raw_to_audio`

**Review by Jefferson111 — COMMENTED** [review]
`raw_to_audio` functionality seems to replaced with just clipping. Quantization to 16 bits and exponential filter are removed. Not sure if its intended.

File: public/externalLibs/sound/sounds.js (hunk 2 above)
Comment (Prearo) [reply public/externalLibs/sound/sounds.js:576]: Quantize was removed as webaudio buffers are stored as float32 arrays.
Removing the exponential decay fade-out was unintended, i'll re-add it back in

File: public/externalLibs/sound/soundToneMatrix.js (hunk 1 above)
Comment (Prearo) [reply public/externalLibs/sound/soundToneMatrix.js:421]: Realized that declaring it broke the library.
will discuss this further offline

File: public/externalLibs/sound/soundToneMatrix.js [hunk 4]
```diff
@@ -357,11 +357,15 @@ function clear_all_timeout() {
 }
 
 // functions from mission 14
+
+const midi_zero_frequency = 8.17579891564371; // base midi frequency
```
Comment (martin-henz) [inline public/externalLibs/sound/soundToneMatrix.js]: Sorry for getting back to this. This number is very odd. The proper definition is A1 which is 440Hz. 
See http://www.cpan.org/modules/by-module/Music/Music-Note-Frequency-0.05.readme

File: public/externalLibs/sound/soundToneMatrix.js [hunk 5]
```diff
@@ -417,7 +421,7 @@ function letter_name_to_frequency(note) {
 }
 
 function midi_note_to_frequency(note) {
-  return 8.1757989156 * Math.pow(2, (note / 12));
+  return midi_zero_frequency * Math.pow(2, (note / 12));
```
Comment (martin-henz) [inline public/externalLibs/sound/soundToneMatrix.js]: formula should change, see above. make it based on A1=440.

File: public/externalLibs/sound/sounds.js [hunk 6]
```diff
@@ -1,24 +1,27 @@
 // Constants
-var FS = 44100; // Standard sampling rate for all problems
+const FS = 44100; // Standard sampling rate for all problems
 
 const fourier_expansion_level = 5; // expansion level for
                                    // square, sawtooth, triangle
 
+const midi_zero_frequency = 8.17579891564371; // base midi frequency
```
Comment (martin-henz) [inline public/externalLibs/sound/sounds.js]: why is this duplicated here?

**Review by martin-henz — CHANGES_REQUESTED** [review]

File: public/externalLibs/sound/soundToneMatrix.js (hunk 4 above)
Comment (Prearo) [reply public/externalLibs/sound/soundToneMatrix.js]: Fixed

**Review by martin-henz — APPROVED** [review]

**Review by LuminousXLB — APPROVED** [review]
